politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go

```diff
@@ -0,0 +1,2976 @@
+			return backend.PluginError{
+				PluginID:  ticketvote.PluginID,
+				ErrorCode: uint32(ticketvote.ErrorCodeVoteOptionsInvalid),
+				ErrorContext: fmt.Sprintf("vote options count got %v, want 2",
+					len(vote.Options)),
+			}
+		}
+		// map[optionID]found
+		options := map[string]bool{
+			ticketvote.VoteOptionIDApprove: false,
+			ticketvote.VoteOptionIDReject:  false,
+		}
+		for _, v := range vote.Options {
+			switch v.ID {
+			case ticketvote.VoteOptionIDApprove:
+				options[v.ID] = true
+			case ticketvote.VoteOptionIDReject:
+				options[v.ID] = true
+			}
+		}
+		missing := make([]string, 0, 2)
+		for k, v := range options {
+			if !v {
+				// Option ID was not found
+				missing = append(missing, k)
+			}
+		}
+		if len(missing) > 0 {
+			return backend.PluginError{
+				PluginID:  ticketvote.PluginID,
+				ErrorCode: uint32(ticketvote.ErrorCodeVoteOptionsInvalid),
+				ErrorContext: fmt.Sprintf("vote option IDs not found: %v",
+					strings.Join(missing, ",")),
+			}
+		}
+	}
+
+	// Verify vote bits are somewhat sane
+	for _, v := range vote.Options {
+		err := voteBitVerify(vote.Options, vote.Mask, v.Bit)
+		if err != nil {
+			return backend.PluginError{
+				PluginID:     ticketvote.PluginID,
+				ErrorCode:    uint32(ticketvote.ErrorCodeVoteBitsInvalid),
+				ErrorContext: err.Error(),
+			}
+		}
+	}
+
+	// Verify parent token
+	switch {
+	case vote.Type == ticketvote.VoteTypeStandard && vote.Parent != "":
+		return backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeVoteParentInvalid),
+			ErrorContext: "parent token should not be provided for a standard vote",
+		}
+	case vote.Type == ticketvote.VoteTypeRunoff:
+		_, err := tokenDecode(vote.Parent)
+		if err != nil {
+			return backend.PluginError{
+				PluginID:     ticketvote.PluginID,
+				ErrorCode:    uint32(ticketvote.ErrorCodeVoteParentInvalid),
+				ErrorContext: fmt.Sprintf("invalid parent %v", vote.Parent),
+			}
+		}
+	}
+
+	return nil
+}
+
+// startStandard starts a standard vote.
+func (p *ticketVotePlugin) startStandard(treeID int64, token []byte, s ticketvote.Start) (*ticketvote.StartReply, error) {
+	// Verify there is only one start details
+	if len(s.Starts) != 1 {
+		return nil, backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeStartDetailsInvalid),
+			ErrorContext: "more than one start details found for standard vote",
+		}
+	}
+	sd := s.Starts[0]
+
+	// Verify token
+	t, err := tokenDecode(sd.Params.Token)
+	if err != nil {
+		return nil, backend.PluginError{
+			PluginID:  ticketvote.PluginID,
+			ErrorCode: uint32(ticketvote.ErrorCodeTokenInvalid),
+		}
+	}
+	if !bytes.Equal(t, token) {
+		return nil, backend.PluginError{
+			PluginID:  ticketvote.PluginID,
+			ErrorCode: uint32(ticketvote.ErrorCodeTokenInvalid),
+			ErrorContext: fmt.Sprintf("plugin payload token does not match "+
+				"route token: got %x, want %x", t, token),
+		}
+	}
```

Comment:
I have seen this code block copy pasted all over this file. We should consolidate the common bits into a single function. This includes sig check and other verification that is common as well.

Reply:
Done.  The same change has also been applied to the comments plugin.

https://github.com/lukebp/politeia/blob/tlog/politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go#L2694